Mismatch between shape Java API

Hi, I’ve finished training the yolo3_darknet53 params and wanted to test it with Java API, however am having this error which I have no idea where it coming from:

     Exception in thread “main” java.lang.IllegalArgumentException: requirement failed:
     Mismatch between shape (100,1) and (1,100,4)
        	at scala.Predef$.require(Predef.scala:224)
        	at org.apache.mxnet.NDArray$$anonfun$9.apply(NDArray.scala:448)
        	at org.apache.mxnet.NDArray$$anonfun$9.apply(NDArray.scala:445)
        	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
        	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
        	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
        	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
        	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        	at org.apache.mxnet.NDArray$.concatenate(NDArray.scala:445)
        	at org.apache.mxnet.module.BaseModule$$anonfun$2.apply(BaseModule.scala:267)
        	at org.apache.mxnet.module.BaseModule$$anonfun$2.apply(BaseModule.scala:267)
        	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
        	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        	at org.apache.mxnet.module.BaseModule.predict(BaseModule.scala:267)
        	at org.apache.mxnet.infer.Predictor$$anonfun$11.apply(Predictor.scala:210)
        	at org.apache.mxnet.infer.Predictor$$anonfun$11.apply(Predictor.scala:210)
        	at org.apache.mxnet.infer.MXNetThreadPoolHandler$$anon$4.call(MXNetHandler.scala:73)
        	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        	at java.lang.Thread.run(Thread.java:748)
    FAILURE: Build failed with an exception.

    * What went wrong:
    Execution failed for task ':ObjectDetectionTutorial.main()'.
    > Process 'command '/usr/lib/jvm/java-8-openjdk-amd64/bin/java'' finished with non-zero exit value 1

    * Try:
    Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

    * Get more help at https://help.gradle.org

    BUILD FAILED in 1s
    2 actionable tasks: 2 executed
    Process 'command '/usr/lib/jvm/java-8-openjdk-amd64/bin/java'' finished with non-zero exit value 1
    10:16:24: Task execution finished 'ObjectDetectionTutorial.main()'.

I only have 1 class and my input image is 512 and pretrained_base=false. Any help would be much appreciated. Thank you in advance.

--- a/scala-package/core/src/main/scala/org/apache/mxnet/module/BaseModule.scala
+++ b/scala-package/core/src/main/scala/org/apache/mxnet/module/BaseModule.scala
@@ -247,6 +247,18 @@ abstract class BaseModule {
 
   /**
    * Run prediction and collect the outputs.
+   * The concatenation process will be like
+   * {{{
+    *   outputBatches = [
+    *      [a1, a2, a3], // batch a
+    *      [b1, b2, b3]  // batch b
+    *   ]
+    *   result = [
+    *     NDArray, // [a1, b1]
+    *     NDArray, // [a2, b2]
+    *     NDArray, // [a3, b3]
+    *   ]
+    * }}}
    * @param evalData
    * @param numBatch Default is -1, indicating running all the batches in the data iterator.
    * @param reset Default is `True`, indicating whether we should reset the data iter before start
@@ -264,7 +276,8 @@ abstract class BaseModule {
           s"in mini-batches (${out.size})." +
       "Maybe bucketing is used?")
     )
-    val concatenatedOutput = outputBatches.map(out => NDArray.concatenate(out))
+    val oBT = outputBatches.transpose
+    val concatenatedOutput = oBT.map(out => NDArray.concatenate(out))
     outputBatches.foreach(_.foreach(_.dispose()))
     concatenatedOutput
   }

--- a/scala-package/core/src/main/scala/org/apache/mxnet/module/BaseModule.scala
+++ b/scala-package/core/src/main/scala/org/apache/mxnet/module/BaseModule.scala
@@ -247,23 +247,23 @@ abstract class BaseModule {
 
   /**
    * Run prediction and collect the outputs.
-   * The concatenation process will be like
-   * {{{
-    *   outputBatches = [
-    *      [a1, a2, a3], // batch a
-    *      [b1, b2, b3]  // batch b
-    *   ]
-    *   result = [
-    *     NDArray, // [a1, b1]
-    *     NDArray, // [a2, b2]
-    *     NDArray, // [a3, b3]
-    *   ]
-    * }}}
-   * @param evalData
+   * @param evalData dataIter to do the Inference
    * @param numBatch Default is -1, indicating running all the batches in the data iterator.
    * @param reset Default is `True`, indicating whether we should reset the data iter before start
    *              doing prediction.
    * @return The return value will be a list `[out1, out2, out3]`.
+   *        The concatenation process will be like
+   *        {{{
+   *            outputBatches = [
+   *              [a1, a2, a3], // batch a
+   *              [b1, b2, b3]  // batch b
+   *            ]
+   *            result = [
+   *              NDArray, // [a1, b1]
+   *              NDArray, // [a2, b2]
+   *              NDArray, // [a3, b3]
+   *            ]
+   *        }}}
    *         Where each element is concatenation of the outputs for all the mini-batches.
    */
   def predict(evalData: DataIter, numBatch: Int = -1, reset: Boolean = true)

--- a/scala-package/core/src/test/scala/org/apache/mxnet/ModuleSuite.scala
+++ b/scala-package/core/src/test/scala/org/apache/mxnet/ModuleSuite.scala
@@ -23,6 +23,34 @@ import org.apache.mxnet.optimizer._
 import org.apache.mxnet.io._
 
 class ModuleSuite extends FunSuite with BeforeAndAfterAll {
+
+  class myModule(symbol : Symbol) extends Module (symbol) {
+    override def predictEveryBatch(evalData: DataIter,
+                                   numBatch: Int = 1, reset: Boolean = true):
+    IndexedSeq[IndexedSeq[NDArray]] = {
+      val data = IndexedSeq(
+        NDArray.ones(Shape(1, 10, 1)),
+        NDArray.ones(Shape(1, 10, 1)),
+        NDArray.ones(Shape(1, 10, 4))
+      )
+      List.fill(numBatch)(data).toIndexedSeq
+    }
+  }
+
+  test("predict") {
+    val sym = Symbol.Variable("data")
+    val mod = new myModule(sym)
+    val dummyIter = new NDArrayIter(IndexedSeq(NDArray.ones(1)))
+    var output = mod.predict(dummyIter, 1)
+    require(output(0).shape == Shape(1, 10, 1))
+    require(output(1).shape == Shape(1, 10, 1))
+    require(output(2).shape == Shape(1, 10, 4))
+    output = mod.predict(dummyIter, 2)
+    require(output(0).shape == Shape(2, 10, 1))
+    require(output(1).shape == Shape(2, 10, 1))
+    require(output(2).shape == Shape(2, 10, 4))
+  }
+
   test ("model dtype") {
     val dType = DType.Float32
     val dShape = Shape(3, 8, 7)

